### PR TITLE
[WIP] Managed Decimal implementation (from CoreRT)

### PR DESCRIPTION
**Pull request by EgorBo**
An attempt to port fully managed Decimal from CoreRT to CoreCLR
Contributes to https://github.com/dotnet/coreclr/issues/18249
Wondering if it passes the tests or not

### CODE REVIEW

**Comment by jkotas** [comment]
@EgorBo Please coordinate this with @pentp . I do not think that the CoreRT implementation is ready to be ported yet. In particular, https://github.com/dotnet/corert/issues/4994 needs to be fixed first.

**Comment by EgorBo** [comment]
@jkotas yeah, it's just a quick attempt 🙂 
need to clean up mscorlib.h 
Surprisingly CoreRT's Decimal works like a charm on mono.

**Comment by EgorBo** [comment]
@jkotas @pentp so all checks are green 🙂 does it mean that the jit-unix bug doesn't reproduce?

File: src/System.Private.CoreLib/src/System/Decimal.Currency.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        //
+        internal static Currency ToCurrency(decimal d)
+        {
+            Currency result = new Currency();
+            FCallToCurrency(ref result, d);
+            return result;
+        }
+
+        [MethodImplAttribute(MethodImplOptions.InternalCall)]
+        private static extern void FCallToCurrency(ref Currency result, decimal d);
```
Comment (pentp) [inline src/System.Private.CoreLib/src/System/Decimal.Currency.cs]: Why is `Decimal.ToCurrency` not using the managed CoreRT variant?

File: src/System.Private.CoreLib/src/System/Number.CoreCLR.cs [hunk 2]
```diff
@@ -2,19 +2,733 @@
...
+    // Decimal, Engineering, Fixed point, General, or Number standard formats
+    // (the C, D, E, F, G, and N format specifiers) are guaranteed to be parseable
+    // by the Parse methods if the NumberStyles.Any style is
+    // specified. Note, however, that the Parse methods do not accept
+    // NaNs or Infinities.
+    //
+    // This class contains only static members and does not need to be serializable 
+
+
     internal static partial class Number
```
Comment (pentp) [inline src/System.Private.CoreLib/src/System/Number.CoreCLR.cs]: There are a lot of unrelated changes for `Double` formatting/parsing - that should be separated from `Decimal` changes if possible.
And code flow direction should be from src/X.cs or X.CoreCLR.cs to shared/X.cs, not from CoreRT to unshared CoreCLR code? The goal is not to copy CoreRT code to CoreCLR but to share it and leave the hopefully minor differences in X.CoreCLR.cs/X.CoreRT.cs files.

File: src/System.Private.CoreLib/src/System/Decimal.cs [hunk 3]
```diff
@@ -56,9 +52,8 @@ namespace System
     // throw an exception. A conversion from float or double to
     // Decimal throws an OverflowException if the value is not within
     // the range of the Decimal type.
-    [StructLayout(LayoutKind.Sequential)]
     [Serializable]
-    [System.Runtime.Versioning.NonVersionable] // This only applies to field layout
+    [StructLayout(LayoutKind.Explicit)]
```
Comment (pentp) [inline src/System.Private.CoreLib/src/System/Decimal.cs]: This breaks unix PInvoke interop as @jkotas already mentioned.

File: src/System.Private.CoreLib/src/System/Number.CoreCLR.cs (hunk 2 above)
Comment (EgorBo) [reply src/System.Private.CoreLib/src/System/Number.CoreCLR.cs]: I agree, initially I just wanted to make sure it compiles and works for CoreCLR in general (because as I said we are going to use it in mono), I'll move it to the shared folder.

File: src/System.Private.CoreLib/src/System/Decimal.cs (hunk 3 above)
Comment (EgorBo) [reply src/System.Private.CoreLib/src/System/Decimal.cs]: Sounds like there should be a test for it

Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Decimal.cs]: #18908

**Comment by EgorBo** [comment]
@jkotas @pentp I've updated my PR:

- Moved `Decimal.cs`, `Decimal.DecCalc.cs`, `Number.(CoreRT/CoreCLR).cs`, `Number.Windows.cs` and `Number.Unix.cs` to shared
- Updated Decimal and Decimal.DecCalc.cs in order to grab https://github.com/dotnet/corert/pull/6091 changes (Decimal is `[StructLayout(LayoutKind.Sequential)]` now)

But there is a failing CoreFX test:

```
07:11:21 
   System.Buffers.Text.Tests.ParserTests.TestParserDecimal(testData: [ParseDecimal '-1e-400',G to [+0.0000000000000000000000000000, scale=28])]) [FAIL]
07:11:21       TestException : This parse attempt [ParseDecimal '-1e-400',G to [+0.0000000000000000000000000000, scale=28])] succeeded as expected but parsed to the wrong value:
07:11:21         Expected: [+0.0000000000000000000000000000, scale=28]
07:11:21         Actual:   [-0.0000000000000000000000000000, scale=28]
```
Probably was introduced by https://github.com/dotnet/corert/pull/6091 ? investigating...

File: src/classlibnative/bcltype/number.cpp [hunk 4]
```diff
@@ -1983,28 +1986,21 @@ STRINGREF NumberToStringFormat(NUMBER* number, STRINGREF str, NUMFMTREF numfmt)
...
 FCIMPLEND
+#else
 
-FCIMPL2(FC_BOOL_RET, COMNumber::NumberBufferToDecimal, NUMBER* number, DECIMAL* value)
+FCIMPL4(int, COMNumber::DoubleToStringUnix, double value, char *format, char *buffer, int bufferLength)
 {
     FCALL_CONTRACT;
-
-    FC_RETURN_BOOL(COMDecimal::NumberToDecimal(number, value) != 0);
+    int result = sprintf_s(buffer, bufferLength, format, value);
```
Comment (jkotas) [inline src/classlibnative/bcltype/number.cpp]: CoreCLR has optimized implementation of DoubleToNumber (look for Grisu3 and Dragon4). It is a lot faster than the `sprintf_s` or `_ecvt_s`. This change is disabling the optimization.

As was pointed out earlier, the formatting/parsing related changes should be looked as separate change.

Comment (EgorBo) [reply src/classlibnative/bcltype/number.cpp]: @jkotas ok, will separate them, however according to my benchmarks (like this:
```
[Benchmark] public decimal Parse5() => decimal.Parse("1024.1024");
etc
```
there is no difference (and even a little bit faster then CoreCLR-master) but I am benchmarking only Windows (probably `sprintf_s` is slower)

**Comment by pentp** [comment]
`-1e-400` parsing difference is caused by https://github.com/dotnet/corert/pull/6091, but the test is too strict: there should be no difference between -0 and +0. I'll look into fixing it.

**Comment by EgorBo** [comment]
@pentp furthermore, I'd say `-0.0000000000000000000000000000` is closer to `-1e-400` then +0... 🙂

**Comment by GSPP** [comment]
Maybe this is a good time to add the features requested in https://github.com/dotnet/corefx/issues/28716.

**Comment by pentp** [comment]
@GSPP It might be a good time to get the new API-s through the API review process. After CoreRT-s decimal code is shared with CoreCLR it should be fairly easy to implement, for example there is already a private `DecCalc.Unscale` function that normalizes a decimal.

**Comment by pentp** [comment]
I created a more focused PR #18948 that does not contain unrelated changes for `Double` formatting/parsing and removes unnecessary `Decimal`+`Currency` native parts.

**Comment by EgorBo** [comment]
ok closing it now
